### PR TITLE
Pause player on internet disconnect and show toast

### LINKED ISSUE #8898: App creates an error report after loss of internet
### Checklist

- [X] I am able to reproduce the bug with the [latest version](https://github.com/TeamNewPipe/NewPipe/releases/latest).
- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have read the [FAQ](https://newpipe.net/FAQ/) and my problem isn't listed.
- [X] I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.
- [X] This issue contains only one bug.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Affected version

v0.23.3

### Steps to reproduce the bug

1. Play video
2. Turn off internet
3. Wait until cached video runs out

### Expected behavior

Stop the video and perhaps show a toast saying something like "No internet connection available"

### Actual behavior

App creates an error report

### Screenshots/Screen recordings

_No response_

## Exception
* __User Action:__ play stream
* __Request:__ Player error[type=ERROR_CODE_IO_NETWORK_CONNECTION_FAILED] occurred while playing https://www.youtube.com/watch?v=uPZOzWFr5KU
* __Content Country:__ US
* __Content Language:__ en-US
* __App Language:__ en_US
* __Service:__ YouTube
* __Version:__ 0.24.0
* __OS:__ Linux Android 12 - 31
<details><summary><b>Crash log </b></summary><p>

```
com.google.android.exoplayer2.ExoPlaybackException: Source error
	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:632)
	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:604)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:226)
	at android.os.Looper.loop(Looper.java:313)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by: com.google.android.exoplayer2.upstream.HttpDataSource$HttpDataSourceException: java.net.UnknownHostException: Unable to resolve host "rr2---sn-1gi7znek.googlevideo.com": No address associated with hostname
	at org.schabi.newpipe.player.datasource.YoutubeHttpDataSource.open(YoutubeHttpDataSource.java:389)
	at com.google.android.exoplayer2.upstream.DefaultDataSource.open(DefaultDataSource.java:258)
	at com.google.android.exoplayer2.upstream.TeeDataSource.open(TeeDataSource.java:52)
	at com.google.android.exoplayer2.upstream.cache.CacheDataSource.openNextSource(CacheDataSource.java:786)
	at com.google.android.exoplayer2.upstream.cache.CacheDataSource.open(CacheDataSource.java:599)
	at com.google.android.exoplayer2.upstream.StatsDataSource.open(StatsDataSource.java:84)
	at com.google.android.exoplayer2.source.chunk.ContainerMediaChunk.load(ContainerMediaChunk.java:124)
	at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:412)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:920)
Caused by: java.net.UnknownHostException: Unable to resolve host "rr2---sn-1gi7znek.googlevideo.com": No address associated with hostname
	at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:124)
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:103)
	at java.net.InetAddress.getAllByName(InetAddress.java:1152)
	at com.android.okhttp.Dns$1.lookup(Dns.java:41)
	at com.android.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:178)
	at com.android.okhttp.internal.http.RouteSelector.nextProxy(RouteSelector.java:144)
	at com.android.okhttp.internal.http.RouteSelector.next(RouteSelector.java:86)
	at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:176)
	at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:128)
	at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:97)
	at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:289)
	at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:232)
	at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:465)
	at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:131)
	at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.connect(DelegatingHttpsURLConnection.java:90)
	at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:30)
	at org.schabi.newpipe.player.datasource.YoutubeHttpDataSource.makeConnection(YoutubeHttpDataSource.java:704)
	at org.schabi.newpipe.player.datasource.YoutubeHttpDataSource.makeConnection(YoutubeHttpDataSource.java:548)
	at org.schabi.newpipe.player.datasource.YoutubeHttpDataSource.open(YoutubeHttpDataSource.java:383)
	... 10 more

```
</details>
<hr>

### Affected Android/Custom ROM version

Android 12

### Affected device model

_No response_

### Additional information

Somewhat related to https://github.com/TeamNewPipe/NewPipe/issues/7651.
 Since the loss of an active internet connection is a somewhat regular occurrence, I think it should be handled better. Especially new users could be put off by an error message + report and it would generally be quality of life improvement if this were to be addressed.

### PR DESCRIPTION
**Pull request by marevm**
#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
- pauses video player when internet disconnects with ERROR_CODE_IO_NETWORK_CONNECTION_FAILED
- disables error notification setting isCatchableException = true
- adds toast method to error util
- modifies play() to prepare() video if video is not prepared

#### Before/After Screenshots/Screen Record
- Before: 
NewPipeError

NewPipeErrorMessage

- After: 
NewPipeNetworkToast

#### Fixes the following issue(s)
- Fixes #8898

#### APK testing
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by opusforlife2** [comment]
@marevm ANU student?

File: app/src/main/java/org/schabi/newpipe/error/ErrorUtil.kt [hunk 1]
```diff
@@ -92,6 +92,14 @@ class ErrorUtil {
...
         }
 
+        /**
+         * Shortcut to calling [Toast.makeText]
+         */
+        @JvmStatic
+        fun showErrorToast(context: Context, message: String) {
+            Toast.makeText(context, message, Toast.LENGTH_LONG).show()
+        }
+
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/error/ErrorUtil.kt:102]: I don't think this is needed, using the `Toast` methods by themselves in classes which need them ensures that you apply the duration show you wanted.

This is a personal opinion, though.

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 2]
```diff
@@ -1423,6 +1423,11 @@ public void onPlayerError(@NonNull final PlaybackException error) {
             case ERROR_CODE_TIMEOUT:
             case ERROR_CODE_IO_UNSPECIFIED:
             case ERROR_CODE_IO_NETWORK_CONNECTION_FAILED:
+                isCatchableException = true;
+                simpleExoPlayer.pause();
+                onPaused();
+                ErrorUtil.showErrorToast(context, "No internet connection available");
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:1429]: You must create a string ressource and reference it here, to allow translation of this message.

File: app/src/main/java/org/schabi/newpipe/player/Player.java [hunk 3]
```diff
@@ -1423,6 +1423,11 @@ public void onPlayerError(@NonNull final PlaybackException error) {
             case ERROR_CODE_TIMEOUT:
             case ERROR_CODE_IO_UNSPECIFIED:
             case ERROR_CODE_IO_NETWORK_CONNECTION_FAILED:
```
Comment (AudricV) [inline app/src/main/java/org/schabi/newpipe/player/Player.java:1425]: As you can see on [the JavaDoc of the error code](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/PlaybackException.html#ERROR_CODE_IO_NETWORK_CONNECTION_FAILED), using this error and show to the user that it doesn't have a connection in this case is not appropriate for all cases.

I recommend checking the network connection using the system APIs / compat/JetPack APIs if there are ones instead.

**Review by AudricV — CHANGES_REQUESTED** [review]

**Comment by opusforlife2** [comment]
> There should be an option 

Or, depending on what the timeout value is, the player could be set to pause just a little before that duration is completed, and keep trying to load until then.

**Review by acrodemocide — COMMENTED** [review]
Do we have any unit tests that should be updated for the new functionality?

**Comment by Stypox** [comment]
> see the player paused every now and then having to click the play button manually

Have you read 8898? The player closes completely at the moment, so this is already better

**Pull request closed without merging**